File: packages/cli/src/commands/console.ts
/**
 * `pilot console <machine>`: a shell on a machine, on a real terminal.
 *
 * It is `machines exec` in terminal mode and nothing more. The same route, the
 * same frames, the same exit verdict: the server grew a `tty` flag on the exec
 * stream rather than a second protocol, so there is one thing to route,
 * authenticate and proxy, and this command opens no socket of its own.
 *
 * Two behaviours are the whole reason it is a separate command:
 *
 *   - the local terminal goes into RAW MODE, so ctrl-c, tab completion and a
 *     full-screen editor reach the remote shell as bytes instead of being
 *     eaten by the local line discipline. That also means a crash which skips
 *     the restore leaves the user with a terminal that no longer echoes, which
 *     is why every exit path here puts it back;
 *   - a window change is forwarded, so `stty size` in the guest keeps agreeing
 *     with the window the user is actually looking at.
 *
 * A session is bound to its socket: closing it kills the shell. There is no
 * detach and reattach, and reaching for one by leaving a socket open would
 * leak a shell, a PTY and two goroutines per abandoned window inside a guest
 * whose whole job is to go idle.
 */

import { Command } from 'commander'

import { clientFromEnv, type GlobalOptions } from '../config.ts'
import { CliError, isJSONMode, note } from '../output.ts'
import { collect, resolveMachine } from '../resolve.ts'
import { execStream, processTerminal, type Terminal } from './machines.ts'

/**
 * A login shell, so the guest's own profile has run before the first prompt.
 *
 * Overridable by anything after `--`, which is how `pilot console box -- bash`
 * or a REPL gets one.
 */
const DEFAULT_ARGV = ['/bin/sh', '-l']

export function createConsoleCommand(term: Terminal = processTerminal): Command {
  return new Command('console')
    .argument('<machine>', 'a machine, by id or name')
    .description('an interactive shell on a machine; everything after -- replaces /bin/sh -l')
    .option('--cwd <dir>', 'working directory in the guest')
    .option('--env <K=V>', 'an environment variable (repeatable)', collect)
    .option('--user <name>', 'run as this user')
    .allowExcessArguments(true)
    .action(async function (this: Command, machine: string) {
      const opts = this.optsWithGlobals() as GlobalOptions & Record<string, unknown>

      // Both refusals come before the fleet is touched. A command that cannot
      // work should say so without a round trip, and without waking a
      // suspended sandbox to tell it.
      //
      // `--json` is checked first because it is the flag the caller typed. A
      // shell session is a stream of terminal bytes with no document at the
      // end of it, so there is no JSON this command could ever print, and
      // half-honouring the flag by printing nothing on stdout would leave a
      // program parsing an empty string as if it were a result.
      if (isJSONMode()) {
        throw new CliError('console is interactive, so it has no --json form', {
          hint: 'for output a program can parse, run pilot machines exec <machine> --json -- <argv...>',
        })
      }
      // Raw mode needs a terminal to put into it, and the initial window comes
      // off stdout. Neither has an answer on a pipe, and a console that
      // half-works off one is worse than one that names the command that does:
      // it connects, draws nothing, and takes the shell down with it on the
      // first stray byte.
      const missing = notATerminal(term)
      if (missing) {
        throw new CliError(`console needs a terminal, and ${missing} is not one`, {
          hint: 'from a script, a pipe or an agent, run pilot machines exec <machine> -- <argv...>',
        })
      }

      const client = clientFromEnv(opts)
      const found = await resolveMachine(client, machine)
      const argv = this.args.slice(1)
      // stderr, never stdout: the session's own bytes are the output here, and
      // a banner mixed into them is a banner in the middle of a `tmux` screen.
      note(`connected to ${found.name} (${found.id}); the shell ends when you leave it`)
      // The remote status, exactly as `machines exec` reports it. Leaving a
      // shell with `exit 3` exits 3.
      process.exitCode = await execStream(
        client,
        found.id,
        argv.length > 0 ? argv : DEFAULT_ARGV,
        { ...opts, tty: true },
        term,
      )
    })
}

/** Which of the two streams is not a terminal, if either. */
function notATerminal(term: Terminal): string | null {
  if (!term.stdin.isTTY) return 'stdin'
  if (!term.stdout.isTTY) return 'stdout'
  return null
}

File: apps/pilot/internal/cli/console.go
package cli

import (
	"os"

	"github.com/spf13/cobra"
	"golang.org/x/term"

	pilots "github.com/vivek7405/pilots/sdks/go"

	"github.com/vivek7405/pilots/cli/internal/out"
)

func newConsoleCmd(env *Env) *cobra.Command {
	c := &cobra.Command{
		Use:     "console [machine] [-- command...]",
		Aliases: []string{"c"},
		Short:   "an interactive shell on a machine",
		Args:    cobra.ArbitraryArgs,
		RunE: func(c *cobra.Command, args []string) error {
			client, err := env.Client()
			if err != nil {
				return err
			}
			name, argv := splitAtDash(c, args)
			m, err := machineArg(c, env, client, name)
			if err != nil {
				return err
			}
			return runConsole(c, env, client, m.ID, argv)
		},
	}
	Describe(c, Doc{
		When: "- `pilot console` -- a shell to work in, on a pseudo-terminal\n" +
			"- `pilot exec`    -- one command, its output back, its exit status yours\n\n" +
			"Reach for console when you want to look around or run several things;\n" +
			"for exec when a script or an agent wants an answer.",
		How: "The window size is forwarded, so full-screen programs lay out\n" +
			"correctly and `stty size` in the guest agrees with your terminal.\n" +
			"Everything after -- replaces the default `/bin/sh -l`. Exiting the\n" +
			"shell ends the session; the machine keeps running.",
		Examples: []string{
			"pilot console scratch",
			"pilot c scratch",
			"pilot console scratch -- bash",
			"pilot console scratch -- htop",
		},
		Related: []string{
			"pilot exec              one command, output back",
			"pilot machines create   creates a machine and opens a console on it",
		},
	})
	return c
}

// newExecCmd is the top-level `pilot exec`, the same as `pilot machines
// exec`. sprite puts `exec (x)` and `console (c)` at the top level because
// they are the two commands an agent runs a hundred times a session; making
// them a level deeper would cost a word each time for no gain.
func newExecCmd(env *Env) *cobra.Command {
	c := newMachinesExecCmd(env)
	c.Use = "exec <machine> -- <command> [args...]"
	return c
}

// runConsole runs argv (or a login shell) on a TTY, with this terminal in raw
// mode so keystrokes reach the guest unedited and the guest's output reaches
// the screen unfiltered.
func runConsole(c *cobra.Command, env *Env, client *pilots.Client, id string, argv []string) error {
	stdinFd, stdoutFd := int(os.Stdin.Fd()), int(os.Stdout.Fd())
	if !term.IsTerminal(stdinFd) || !term.IsTerminal(stdoutFd) {
		return out.Failf("run it from a terminal, or use `pilot exec` for a command without one",
			"a console needs stdin and stdout to be a terminal")
	}
	if len(argv) == 0 {
		argv = []string{"/bin/sh", "-l"}
	}

	cols, rows, err := term.GetSize(stdoutFd)
	if err != nil {
		cols, rows = 80, 24
	}
	stream, err := client.Machines.ExecStream(c.Context(), id, argv, pilots.ExecStreamOptions{
		Stdin: true, TTY: true, Cols: uint16(cols), Rows: uint16(rows),
	})
	if err != nil {
		return err
	}
	// Raw mode, window size, ctrl-\ to detach: the same loop `pilot attach`
	// runs, because a console IS a session from its first frame.
	return driveTerminal(env, stream, stdinFd, stdoutFd, "")
}
